internal/compiler/fileloader.go

```diff
@@ -221,9 +224,31 @@ func (t *parseTask) start(loader *fileLoader) {
 	})
 }
 
+func (p *fileLoader) CacheSourceFileMetaData(path tspath.Path) {
+	p.sourceFileMetaDatasMutex.Lock()
+	defer p.sourceFileMetaDatasMutex.Unlock()
+	if _, ok := p.sourceFileMetaDatas[path]; ok {
+		return
+	}
+
+	if p.sourceFileMetaDatas == nil {
+		p.sourceFileMetaDatas = make(map[tspath.Path]*ast.SourceFileMetaData)
+	}
+
+	packageJsonType := p.resolver.GetPackageJsonTypeIfApplicable(string(path))
+	impliedNodeFormat := ast.GetImpliedNodeFormatForFile(string(path), packageJsonType)
+	metadata := &ast.SourceFileMetaData{
+		PackageJsonType:   packageJsonType,
+		ImpliedNodeFormat: impliedNodeFormat,
+	}
+
+	p.sourceFileMetaDatas[path] = metadata
+}
+
 func (p *fileLoader) parseSourceFile(fileName string) *ast.SourceFile {
 	path := tspath.ToPath(fileName, p.host.GetCurrentDirectory(), p.host.FS().UseCaseSensitiveFileNames())
 	sourceFile := p.host.GetSourceFile(fileName, path, p.compilerOptions.GetEmitScriptTarget())
+	p.CacheSourceFileMetaData(path)
```

Comment:
Nit, but `CacheSourceFileMetaData` shouldn't be exported, and I would probably rename it to be like `processSourceFileMetaData` or `loadSourceFileMetaData` or something (everything in this file is "caching" information for later, in a way, it's the loader).